### PR TITLE
Fix send code validation

### PR DESCRIPTION
**Pull request by teneeto**
### Details

This PR fixes issue with sending email validation link. When a user triggers email validation from profile, the magic sign-in code isn't sent immediately till you have to trigger `Didn't receive a magic code?`. This PR is raise to fix a reverted code merge https://github.com/Expensify/App/pull/46846

### Fixed Issues

$: [41330](https://github.com/Expensify/App/issues/41330)
PROPOSAL: [issuecomment](https://github.com/Expensify/App/issues/41330#issuecomment-2275066832)

### Tests

**Verifying newly added account**
1. Make sure you are Logged out of both NewDot and OldDot
2. Navigate to OldDot https://staging.expensify.com/
3. In the login screen enter a new gmail account (make sure it's a gmail account as my work email didn't work)
4. Verify you're navigated to staging NewDot
5. Navigate to account Settings > profile > Contact methods
6. Click on the email
7. Verify you're prompted for the magic code to validate the account
8. Verify the account receives a magic code to their email inbox
9. Enter the magic code in the field
10. Click on Verify

**Adding a Secondary Contact**
1. add a new contact method
2. clicking on "add" button should take you to validate code form
3. Clicking on a contact in the contact methods should trigger a validate code request

**Expectation**: The validation link is sent first time, and user is not logged out when magic code is entered.

- [x] Verify that no errors appear in the JS console

### Offline tests

**Verifying newly added account**
1. Make sure you are Logged out of both NewDot and OldDot
2. Navigate to OldDot https://staging.expensify.com/
3. In the login screen enter a new gmail account (make sure it's a gmail account as my work email didn't work)
4. Verify you're navigated to staging NewDot
5. Navigate to account Settings > profile > Contact methods
6. Click on the email
7. Verify you're prompted for the magic code to validate the account
8. Verify the account receives a magic code to their email inbox
9. Enter the magic code in the field
10. Click on Verify

**Adding a Secondary Contact**
1. add a new contact method
2. clicking on "add" button should take you to validate code form
3. Clicking on a contact in the contact methods should trigger a validate code request

**Expectation**: The validation link is sent first time, and user is not logged out when magic code is entered.
### QA Steps

**Verifying newly added account**
1. Make sure you are Logged out of both NewDot and OldDot
2. Navigate to OldDot https://staging.expensify.com/
3. In the login screen enter a new gmail account (make sure it's a gmail account as my work email didn't work)
4. Verify you're navigated to staging NewDot
5. Navigate to account Settings > profile > Contact methods
6. Click on the email
7. Verify you're prompted for the magic code to validate the account
8. Verify the account receives a magic code to their email inbox
9. Enter the magic code in the field
10. Click on Verify

**Adding a Secondary Contact**
1. add a new contact method
2. clicking on "add" button should take you to validate code form
3. Clicking on a contact in the contact methods should trigger a validate code request

**Expectation**: The validation link is sent first time, and user is not logged out when magic code is entered.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/b187becc-5972-4a2c-915d-01811dab0d45

</details>
<details>
<summary>iOS: Native</summary>

</details>
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by teneeto** [comment]
Hi @c3024, can we check for the update here.

**Review by c3024 — COMMENTED** [review]
Automation did not work. Commenting so that it appears on my K2.

**Review by c3024 — CHANGES_REQUESTED** [review]
Based on the clarification for case (2) [here](https://github.com/Expensify/App/issues/41330#issuecomment-2319063902),

we need changes here,
https://github.com/Expensify/App/blob/fc849dda35628b8c6c8d6edfa66dd31b257e8e7a/src/libs/actions/User.ts#L347

```js
-     Navigation.goBack(ROUTES.SETTINGS_CONTACT_METHODS.getRoute(backTo));
+     Navigation.navigate(ROUTES.SETTINGS_CONTACT_METHOD_DETAILS.getRoute(contactMethod));
```
and here,
https://github.com/Expensify/App/blob/fc849dda35628b8c6c8d6edfa66dd31b257e8e7a/src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.tsx#L152
```js
-         Navigation.goBack(ROUTES.SETTINGS_CONTACT_METHODS.getRoute(backTo));
+    Navigation.navigate(ROUTES.SETTINGS_CONTACT_METHODS.getRoute(backTo));
```

**Comment by teneeto** [comment]
Hi @c3024, let's check this again.

**Comment by c3024** [comment]
The first change mentioned [here](https://github.com/Expensify/App/pull/47772#pullrequestreview-2269215113) for `User.ts` is required for sending code correctly for secondary contacts. This is not included in the PR. Otherwise the API request is made for sending code for the primary contact.

Here is the video showing this bug.


https://github.com/user-attachments/assets/4418c44f-3ae6-486c-afea-47fe9dd044c2

**Comment by teneeto** [comment]
Hi @c3024, I haven't added that piece yet, seeing that there was a recent code change which removed navigation completely: https://github.com/Expensify/App/pull/48320/files#r1738644645

I'm still trying to figure out why that was done and what next to do. As it stands, adding that piece gives this flickering navigation behaviour as shown in the screen record

https://github.com/user-attachments/assets/7b62a2a4-5dfe-4b3b-995f-3a4397f1ddf4

**Comment by c3024** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Review by c3024 — APPROVED** [review]
LGTM!

**Comment by shawnborton** [comment]
Why does the UI look differently between these two screens?
CleanShot 2024-09-19 at 14 45 37@2x

CleanShot 2024-09-19 at 14 46 00@2x

**Comment by dannymcclain** [comment]
Also this is feeling verrrrrry tight to me:

CleanShot 2024-09-19 at 08 23 23@2x

Can we try `20px` between the number inputs and the `Didn't receive a magic code?` message? And then, is there a reason the `Verify` button isn't at the bottom like our standard "footer buttons"?

**Comment by teneeto** [comment]
> Why does the UI look different between these two screens? CleanShot 2024-09-19 at 14 45 37@2x
> 
> CleanShot 2024-09-19 at 14 46 00@2x

Hi @shawnborton, Although I didn't make any UI changes here. I can make changes if you could propose what the  UI should look like.

**Comment by teneeto** [comment]
Hi @c3024, what's the plan with this PR, anything more to push for?

**Comment by nkuoch** [comment]
Let me know when it's ready for final review

**Comment by teneeto** [comment]
@nkuoch yes this is ready for review.

**Review by nkuoch — DISMISSED** [review]

**Comment by nkuoch** [comment]
cc @shawnborton for https://github.com/Expensify/App/pull/47772#issuecomment-2376063088

**Comment by shawnborton** [comment]
Sorry - I've been OOO for the past 3 weeks so just catching up now. Is the original comment updated with the latest screenshots of this PR?

**Comment by teneeto** [comment]
> Sorry - I've been OOO for the past 3 weeks so just catching up now. Is the original comment updated with the latest screenshots of this PR?

Hi @shawnborton, welcome back from ooo.  I had worked on this issue for code validation and secondary email validation flow; I would need more direction on what should be done on that UI and clarification on what to do according to the original comment.

**Comment by teneeto** [comment]
>Why does the UI look differently between these two screens?
1. You mentioned that, and I'm wondering what the actual UI between those screens should be (I couldn't get the difference you mentioned)
2. @dannymcclain mentioned why the verify button isn't at the bottom like other standard footer buttons - should I make it that way?

I will pick this up once I have clarification and green light on those.

**Comment by shawnborton** [comment]
> mentioned why the verify button isn't at the bottom like other standard footer buttons - should I make it that way?

Yes! 

In this screen, the message above the form has a green dot pattern:
image

I kind of think we can just kill that and use a standard message like we have here:
image

But one way to think about this - if you basically go to create a new account and don't validate it, then go to add a bank account, let's just reuse the same kind of screen/form that you see here: 
CleanShot 2024-10-14 at 21 18 45@2x

**Comment by teneeto** [comment]
Lovely, I'll start fixes ASAP @shawnborton 

One question and clarification: The screen with the `delete button` should be left that way, right? We don't have to push the `verify button` to the footer since the `delete button` is below it, right?

**Comment by dannymcclain** [comment]
For that case I think we'd do it like this:

image

But happy for Shawn to weigh in too!

**Comment by shawnborton** [comment]
Yup, I dig it!

**Comment by teneeto** [comment]
@shawnborton @dannymcclain, this should be a new look for the validation flow.

https://github.com/user-attachments/assets/3bb73362-f396-4252-8d17-787e65b982ad

____________
One question is, since we removed the dot indicator (green) for contacts that require verification, should we also remove the dot indicator (red) for failed contacts (contacts that cannot be added because they already exist, etc.)? Please take a look at the video record below.

https://github.com/user-attachments/assets/82e8a158-c0ba-48a0-a48d-6d2b4084c2ea

**Comment by shawnborton** [comment]
Oh sorry, I don't think we should remove the green dot from this screen here: 
CleanShot 2024-10-17 at 11 26 49@2x

I just think we should remove it from the message on the subsequent screen. 

So that being said, I think we should keep the green and red dot indicators on the Contact methods page.

**Comment by teneeto** [comment]
Hey @shawnborton, that's precisely my question 😄. The red dot indicator hasn't been removed from the contact methods; I am confirming if it should be removed from the details page. Which screen is the next?

**Comment by shawnborton** [comment]
Ah sorry, I would say no, do not remove it from the subsequent page. We need to show an error message there.

**Comment by teneeto** [comment]
> Ah sorry, I would say no, do not remove it from the subsequent page. We need to show an error message there.

For clarity and confirmation, 
1. I should remove the green dot on the subsequent page.
2. I should not remove the red dot on the subsequent page

Are those scenarios correct? @shawnborton

**Comment by shawnborton** [comment]
Correct!

**Comment by teneeto** [comment]
@nkuoch @c3024 This is ready for another review. Let's check.

**Comment by c3024** [comment]
"Remove" button is missing in the secondary contact details page.

Alignment of "Set as default" button is too close to the left edge.


Dev and Prod:

https://github.com/user-attachments/assets/c90a778c-635e-41ad-8831-429631437f47

Staging:

Screenshot 2024-10-19 at 12 43 22 PM

**Comment by teneeto** [comment]
@c3024, thanks for catching that. I will fix it and push back.

**Comment by teneeto** [comment]
@c3024, one more time, let's test again 🙏.

**Comment by c3024** [comment]
Shouldn't clicking on "Didn't receive a magic code" trigger a `RequestContactMethodValidateCode` request?


https://github.com/user-attachments/assets/f452c618-a2a5-4e55-a363-9b7b0bb86943


Also, please update the "Tests" section of the PR author template to reflect the steps that appropriately test the changes of this PR.

**Comment by teneeto** [comment]
> Shouldn't clicking on "Didn't receive a magic code" trigger a `RequestContactMethodValidateCode` request?
> 
>  resendValidateCode.mp4 
> Also, please update the "Tests" section of the PR author template to reflect the steps that appropriately test the changes of this PR.

Other PRs constantly update this. That was working before; it's hard to tell what decisions were made in other PRs. I didn't consider that to be within the scope of this issue, but I can check again.

**Comment by c3024** [comment]
Oh, sorry! That issue is there on `main` as well. I will review again. Thanks!

Can you please update the "Tests" to match the behaviour we have now like 
1. after adding a secondary contact we should go to the validate code form and 
2. clicking on a contact in the Contact Methods page triggers a ValidateCode request.

**Comment by teneeto** [comment]
Updated @c3024

**Comment by c3024** [comment]
@teneeto

I found a couple of issues:

	1.	Our last PR was reverted because two magic codes were triggered when a new contact method was added. This is still happening; a code is triggered from both the front-end and the backend.
	2.	When we remove a contact method from the magic code page, we should be navigated back to the contact methods page, in my opinion.

**Comment by teneeto** [comment]
The case is slightly different from the reason for the reversal. We initially had multiple validations in the user mailbox, and as we speak, no single piece of code modifies or triggers a validation code request in this PR, and there have been lots of other PR lately rapidly making refactoring or changes to this validation part. Let's open another discussion or issue to look into that.
We still need to revisit why we are fixing this issue.

**Comment by c3024** [comment]
Based on all comments so far, I think, at minimum, these tests should pass without breaking any existing functionality on main with this PR.

Test 1:

	1.	Navigate to “Contact Methods Page.”
	2.	Click on a “Contact Method.”
	3.	A validation code should be sent immediately without needing to click “Resend code.”

Test 2:

	1.	Navigate to “Contact Methods Page.”
	2.	Click on “Add a new Contact Method.”
	3.	Input an email > Add.
	4.	A magic code should be received via email for the primary contact without needing to click “Resend code.”
	5.	Validate this primary contact using the magic code.
	6.	Once validated, the secondary contact is added, and the backend sends a magic code for this secondary contact’s validation. The user should be navigated to the validation code form for the secondary contact.
	7.	No duplicate magic code should be received on the secondary contact’s email.
	8.	Validating the secondary contact with the magic code should navigate the user back to the “Contact Methods Page.”

Test 3:

	1.	Navigate to “Contact Methods Page.”
	2.	Click on “Add a new Contact Method.”
	3.	Input an email > Add.
	4.	A magic code should be received via email for the primary contact without needing to click “Resend code.”
	5.	Validate this primary contact with the magic code.
	6.	Once validated, the secondary contact is added, and the backend sends a magic code for the secondary contact’s validation. The user should be navigated to the validation code form for the secondary contact.
	7.	No duplicate magic code should be received on the secondary contact’s email.
	8.	Clicking “Remove” should delete the newly added contact method, and the user should be navigated back to the “Contact Methods Page.”

**Comment by c3024** [comment]
Test 1 above should definitely pass with this PR, and I don’t think we can make changes solely for Test 1 without affecting the behaviours in Test 2 and Test 3. If we can pass Test 1 while retaining the other flows in Test 2 and 3 as they are on `main`, I think that is fine.

**Comment by teneeto** [comment]
All should pass now; I did another merge conflict, hoping no new changes. Can we check.

**Comment by c3024** [comment]
Test 1 works well even on main.

With this PR, Test 2 fails at step 7. Two magic codes, one from the backend and one from the frontend, are received.

**Comment by teneeto** [comment]
Ok let me have a look at step 2.

**Comment by teneeto** [comment]
@c3024 can we check again?

File: src/pages/settings/Profile/Contacts/NewContactMethodPage.tsx
Comment (c3024) [inline src/pages/settings/Profile/Contacts/NewContactMethodPage.tsx:63]: With this change, even when the user submits an incorrect magic code while validating the primary contact, it navigates to the secondary contact method page. I think the validation code form for the primary contact should remain visible until the backend confirms the code entered is correct. If the backend indicates the code is incorrect, an error message should be displayed on the primary contact form itself. Only after the backend confirms the code’s validity should the user be navigated to the validation code form for the newly added contact method.


https://github.com/user-attachments/assets/f95167df-50bd-4dee-8e84-11fb341dbd33

**Review by c3024 — CHANGES_REQUESTED** [review]

File: src/pages/settings/Profile/Contacts/NewContactMethodPage.tsx
Comment (teneeto) [reply src/pages/settings/Profile/Contacts/NewContactMethodPage.tsx:63]: I think there was a check for that, which I removed. I'll add it back

**Comment by teneeto** [comment]
@c3024 can we check this one again

**Comment by c3024** [comment]
When we validate the primary contact the secondary contact should be added to the contact methods.

Earlier, on validating the primary contact with magic code in the new contact method flow, it navigated to `ContactMethodsPage`. 

Now, it navigates to the `ContactMethodDetails` page and it is possible to go back without validating this secondary contact. Going back again on the primary contact validation code page removes this newly added but not yet validated secondary contact locally 

https://github.com/Expensify/App/blob/c18812ae3cac59025da0f6c8657750dfcb779a65/src/pages/settings/Profile/Contacts/NewContactMethodPage.tsx#L168

and this does not appear in `ContactMethodsPage`.

On this branch vs staging

https://github.com/user-attachments/assets/7f5f950e-f843-41c4-8ba8-7593749d261c

The added secondary contact method, however, was written to backend so it shows up if we log out and login.


https://github.com/user-attachments/assets/7594962a-933a-43f0-8c64-ced722df6eb6

**Comment by waterim** [comment]
@c3024 Should work fine now

**Comment by waterim** [comment]
@c3024 bump

**Comment by waterim** [comment]
@c3024 bump

**Comment by c3024** [comment]
Thanks for the bump. I will check and update!

**Comment by c3024** [comment]
Magic code is sent twice when adding a secondary contact.

Like mentioned [here](https://github.com/Expensify/App/pull/47772#issuecomment-2447302832) in Step 7 in Test 2, there should be no duplicate magic code emails when adding a secondary contact.

https://github.com/user-attachments/assets/bfbdd1cb-e8b8-45c0-a48b-b4a910d9c7d3

**Comment by waterim** [comment]
Ahh, okay, didn't see those steps

**Comment by waterim** [comment]
update: found a root, trying to fix it without breaking other parts

**Comment by waterim** [comment]
And also on staging Im getting the same issue with double emails
image

**Comment by c3024** [comment]
@waterim

On staging:
	1.	Add a contact method.
	2.	Validate the primary contact.
	3.	Navigates to the Contact Method Page.

Between steps 2 and 3, a magic code was sent by the backend. Then, on clicking the contact method, another magic code was sent due to the frontend request.

On this branch:
	1.	Add a contact method.
	2.	Validate the primary contact.
	3.	Navigates to the validation modal for the secondary contact.

Between steps 2 and 3, a magic code was sent by the backend. When the secondary contact method validation modal opens, another magic code was sent due to the frontend request.

History of this issue:
	1.	There was an issue where clicking on a contact method did not send a magic code immediately.
	2.	That was fixed with PR #45395.
	3.	This was reverted because of [this](https://github.com/Expensify/App/issues/41330#issuecomment-2312618131).
	4.	[This](https://github.com/Expensify/App/issues/41330#issuecomment-2319063902) solution addresses the prevention of duplicate codes.
	5.	The present PR tried to tackle both issues.
	6.	Meanwhile, the issue of not sending a magic code on clicking a contact method seems to have been fixed by another PR.
	7.	The remaining issue is to fix the duplicate code issue without breaking any other functionality. Based on the previous reviews, I found [these](https://github.com/Expensify/App/pull/47772#issuecomment-2447302832) conditions that usually broke due to changes in this PR and suggested relevant tests.

**Comment by waterim** [comment]
@c3024 Didnt you get the same not found on this branch after I merged main?

**Comment by waterim** [comment]
Thank you for the detailed history, it helps a lot!

**Comment by c3024** [comment]
Yes, I am also seeing the same Not Found after validating the primary contact when adding a secondary contact.

**Comment by waterim** [comment]
Okay, finally I found that this [change ](https://github.com/Expensify/App/pull/53742) caused this not-found, as now contactMethod is setting to null

To be honest Im a bit confused with all these changes, this flow is changing all the time and each merge breaking some part of the logic

Initial issue for this PR was also closed yesterday, do we still want this PR to proceed?
As far as I understand this PR was about to fix the code validation, but right now it looks like we changed the flow fully, but during this PR is ongoing there are a lot of changes from outside coming to this flow and it looks like everything works perfectly without this changes from here.


Can you please describe the desired/expected solution from this PR?

**Comment by c3024** [comment]
Yes, I think we can close this PR. If necessary, a new issue can be created to fix the duplicate codes issue.

**Pull request closed without merging**
